### ISSUE TITLE
Custom matplotlib modules can't work with 2.1 AND 2.2

### ISSUE THREAD

**Issue by remram44** [VisTrails/VisTrails#1067]
See also: #706

In 2.1, matplotlib actually perform operations, (wrongly) assuming that the current figure is always automatically the one they are supposed to work on.

In 2.2, to fix various problems (#396, #672), matplotlib modules pass down functions that perform these operations, so that MplFigure can run them on the actual figure (even several times on different figures, if requested).

The problem is that the code, execution model, and type of objects passed on connections is completely different. Custom code extending `MplPlot` will not work on both 2.1 and 2.2. It might also break in the future if we finally move to a sane way of plotting (if/when matplotlib provides one).

Somehow there should be a compatibility layer allowing one to write a matplotlib module that works on all versions.

**Comment by remram44** [comment]
We can either add a class that would be compatible, meant for users to extend, or make the new MplPlot be compatible with 2.1 (which would probably mean changing the way it works in 2.2).
```python
import matplotlib.pyplot
from vistrails.packages.matplotlib.bases import MplPlot, MplSource

class MplCustomPlot(MplPlot):
    if hasattr(MplSource, 'plot_figure'):
        # 2.2
        _output_ports = [('value', '(MplCustomPlot)')]
        def compute(self):
            self.set_output('value', lambda figure: self.plot())
    else:
        # 2.1
        _output_ports = [('self', MplPlot, {'optional': True}),
                         ('value', '(MplCustomPlot)')]
        def compute(self):
            self.setResult('value', self)
            self.plot()

    def plot(self):
        raise NotImplementedError("Subclasses of MplCustomPlot must implement "
                                  "plot()")

class MplCompatTest(MplCustomPlot):
    _input_ports = [('x', '(basic:List)')]

    def plot(self):
        x = self.getInputFromPort('x')
        print repr(x)
        matplotlib.pyplot.hist(x)

_modules = [(MplCustomPlot, {'abstract': True}),
            MplCompatTest]
```

**Comment by remram44** [comment]
Decision after meeting: don't try to do magic in the matplotlib package to try and restore compatibility. Instead, provide the MplCustomPlot class.
I will look into ALPS, see if I can get the plots to work.

**Comment by vistrails-dev** [comment]
Indeed, I always have to manually upgrade the subworkflows with every VisTrails version. I just open the XML files, move one of the modules, and save the file again.

**Comment by remram44** [comment]
Ok, I made the changes on branch `alps-mpl`. The example I have gives me a blank plot though (`3bcbed00b2130ab77c803539bc78dfc6.xml`: the PreparePlot returns a dict with `'data': []`), what is a good example to try this on?

**Comment by vistrails-dev** [comment]
Thank you! I pulled the changes and it now works after I modified one of my subworkflows. I can create plots again. I will next try to bump the version number of the ALPS package (which as you realized always causes issues) and will then build a new version.

Matthias
